### PR TITLE
5.4.x: PTP_KVM support for arm/arm64

### PR DESCRIPTION
**Pull request by damon-kwok**
The enablement of ptp_kvm for arm is completed, see [[1]](https://lore.kernel.org/kvmarm/20210330145430.996981-1-maz@kernel.org/).
It's for `kernel 5.13.x`, so we have to apply private patch to enable it in `kernel 5.4`.
The `ptp_kvm` can offer the capability of time sync in `kata` even there
is no network available and higher precision than time sync
service depend on network.
I saw that Archana Shinde deleted the ARM patch of `ptp_kvm`，
 because this patch caused many error in the `x86` build [[2]](kata-containers/packaging@82c6318#diff-dedd9a2bb757a0d2f27b981d37edf70b563e94735bbe76508e02e4f4aff93575).
I refactored the patch to fix the build problem on `x86` and tried to reactivate it.

In order for the `ptp_kvm` module to work on all kernel versions, I did some work. 
I merged the code from  `kernel 5.13`, and made patches for different versions.
[patch-1](https://github.com/kata-containers/kata-containers/pull/2105) for `kernel 4.19.x LTS`
[patch-2](https://github.com/kata-containers/kata-containers/pull/2058) for `kernel 5.4.x LTS`
[patch-3](https://github.com/kata-containers/kata-containers/pull/2105) for `kernel 5.10.x LTS`

Test:
|HOST  |  GUEST |  RESULT|
| --- | -------|---- |
|4.19   |  4.19   |  pass|
|5.4   |   5.4    |  pass|
|5.10   |  5.10   |  pass|
|4.19   |  5.4    |  pass|
|4.19  |   5.10   |  pass|
|5.4   |   4.19   |  pass|
|5.4    |  5.10   |  pass|
|5.10  |   5.4    |  pass|
|5.10  |   4.19   |  pass|
|5.12  |   4.19   |  pass|
|4.19  |   5.12   |  pass|

See:
```
[root@localhost ~]# uname -sm
Linux aarch64
[root@localhost ~]# modprobe ptp_kvm
[root@localhost ~]# chronyc sources
210 Number of sources = 1
MS Name/IP address         Stratum Poll Reach LastRx Last sample               
===============================================================================
#* PHC0                          0   2   377     4    -13ns[  -61ns] +/-  181ns
```
---
- [1] https://lore.kernel.org/kvmarm/20210330145430.996981-1-maz@kernel.org/
- [2] https://github.com/kata-containers/packaging/commit/82c6318b9dc78feab07775fd92a2a9578cbf55fe#diff-dedd9a2bb757a0d2f27b981d37edf70b563e94735bbe76508e02e4f4aff93575

### CODE REVIEW

**Comment by jodh-intel** [comment]
Thanks @damon-kwok!

Please could you add a couple of extra details to the commit message:

- A reference to the PR that removed the original versions of the patchset
  (https://github.com/kata-containers/packaging/commit/82c6318b9dc78feab07775fd92a2a9578cbf55fe#diff-dedd9a2bb757a0d2f27b981d37edf70b563e94735bbe76508e02e4f4aff93575)

- Ideally also add a canonical URL for these new patches (are they on https://patchwork.kernel.org ?)

Also, let's get input from @jongwu, @Pennyzct and @justin-he.

**Comment by jongwu** [comment]
Hello @damon-kwok -,  thanks for your grate job! I'm glad to see that the ptp_kvm module on arm is used by someone. Even the current kernel for kata has upgrade to 5.10, these old patches are also meaningful. Thanks for adding them back.
As @jodh-intel say, you can reference the link [here](https://lore.kernel.org/kvmarm/20210330145430.996981-1-maz@kernel.org/) 

Thanks @jodh-intel -, I'm ok with this PR since it has been tested by @damon-kwok .

**Comment by damon-kwok** [comment]
Hi, @jodh-intel I modified the details.

**Comment by damon-kwok** [comment]
https://github.com/kata-containers/kata-containers/pull/2063#issuecomment-865797304

Don't Merge.

**Comment by damon-kwok** [comment]
@jodh-intel @jongwu @Pennyzct  @justin-he
In order for the `ptp_kvm` module to work on all kernel versions, I did some work. 
I merged the code from  `kernel 5.13`, and made patches for different versions.
#2058 #2063 #2105

File: tools/packaging/kernel/patches/5.4.x/0003-5.4.x-PTP_KVM-support-for-arm-arm64-in-Kata.patch [hunk 1]
```diff
@@ -129,145 +246,105 @@ index 84f4ff351..87a7dc18b 100644
...
+index 0517272..f647a2f 100644
 --- a/drivers/ptp/Kconfig
 +++ b/drivers/ptp/Kconfig
 @@ -110,7 +110,7 @@ config PTP_1588_CLOCK_PCH
  config PTP_1588_CLOCK_KVM
  	tristate "KVM virtual PTP clock"
  	depends on PTP_1588_CLOCK
 -	depends on KVM_GUEST && X86
-+	depends on KVM_GUEST && X86 || ARM64 && ARM_ARCH_TIMER
++	depends on X86 || (HAVE_ARM_SMCCC_DISCOVERY && ARM_ARCH_TIMER)
```
Comment (jongwu) [inline tools/packaging/kernel/patches/5.4.x/0003-5.4.x-PTP_KVM-support-for-arm-arm64-in-Kata.patch:257]: please add "KVM_GUEST" here.

**Pull request closed without merging**